docs/flip/2434-general-metadata.md

```diff
@@ -0,0 +1,143 @@
+4. No need to lift metadata collections seperately.
+
+
+Disadvantages:
+1. Handling boxed values requires the relatively new ``is_leaf=`` syntax which users might not be familiar with. Although users will
+   probably call Flax provided utils that handle the low-level tree_map calls in most cases.
+3. Custom Pytree nodes have a small runtime overhead. It's hard to observe this in practise because JAX calls are async.
+
+
+### Init syntax
+
+
+Boxes can be created directly by the init function of a variable. Therefore, we propose to create metadata using higher-order initializers.
+The main advantage of this is that we can decouple metadata handling compltely from the Module definition. Also, most Modules already overwrite
+attributes to override the default initialzers so users can add metadata to existing Modules without requiring any code changes.
+
+To illustrate this, let's consider a metadata class that keeps track of PartitionSpecs used by ``pjit``:
+
+```python
+class Partitioned(flax.struct.PyTreeNode, AxisMetadata):
+  value: Any
+  named_axes: Tuple[Optional[str], ...]
+  
+  def add_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
+    axis_name = self._get_partition_name(params)
+    names = list(self.names)
+    names.insert(index, axis_name)
+    return self.replace(names=tuple(names))
+
+  def remove_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
+    axis_name = self._get_partition_name(params)
+    names = list(self.names)
+    assert names.pop(index) == axis_name
+    return self.replace(names=tuple(names))
+
+def with_partitioning(init_fn, names):
+  def wrapper(*args, **kwargs):
+    return Partitioned(init_fn(*args, **kwargs), names)
+  return wrapper
+```
+
+Here we also defined a small utility called ``with_partitioning`` that we can use to wrap existing initialzers to add metadata: 
+
+
+```python
+# init kernel with lecun normal and split the output features over the data axis
+partitioned_dense = nn.Dense(features, kernel_init=with_partitioning(nn.initializers.lecun_normal, (None, "data")))
+```
+
+
+### Unbox syntax
+
+
+Metadata typically doesn't need to be handled by Modules directly. Therefore, we propose to make Modules agnostic to Metadata boxes by default.
+The ``unbox`` method can be used to unpack a variable such that only the original JAX arrays remain. Users can manually call unbox but to make
+sure Module classes don't have to call it everywhere we add an unbox keyword arg to variable returning APIs (e.g.: ``.param``, ``.variable``, ``.get_variable``).
+The keyword arg ``unbox`` will default to ``True`` such that a Modules are metadata agnostic by default. This also means existing Modules will be backward compatible
+with the new API.
+
+```python
+kernel = self.param("kernel", self.kernel_init, shape)  # No AxisMetadata instances
+kernel_box = self.get_variable("param", "kernel", unbox=False)  # AxisMetadata boxes are preserved
+```
+
+
+### Lift syntax
```

Comment:
to make the FLIP a bit more complete, it would be interesting to extend the `Partitioned` example a bit to show the entire lifecycle of the (un)boxing, i.e. lifting the `partitioned_dense`, showing the variable pytree etc.

Reply:
Very nice illustration.